slang.h

```diff
@@ -450,6 +450,178 @@ convention for interface methods.
+the `PrimaryStruct`. This also provides a mechanism such that a `PrimaryStruct` derived type does not need to include all the fields that will
+every be needed. Cross cutting aspects can have their own uniquely identified structs.
+
+As a mechanism to provide extensibility without having to modify a struct or to allow optional and/or additional information through
+ExtensionStruct set via the `exts` and `extsCount` members.
+
+A type that is ABI compatible with this mechanism must start with exactly the same fields as specified.
+When using extensions, the exts and extsCount should be set. `exts` points to the extension ids (which must be at the start of the types)
+to provide some kind of type safely without requiring inheritance.
+*/
+
+#define SLANG_EXTENSION_TAGGED_STRUCT_IMPL(TYPE_NAME, CATEGORY, TYPE_ID, MAJOR, MINOR) \
+    typedef TYPE_NAME ThisType; \
+    typedef slang::ExtensionTag Tag; \
+    static const slang::StructTag kStructTag = SLANG_MAKE_EXTENSION_STRUCT_TAG(CATEGORY, TYPE_ID, MAJOR, MINOR); \
+    slang::StructTag structTag = kStructTag; \
+    slang::StructSize structSize = slang::StructSize(sizeof(ThisType));
+
+#define SLANG_PRIMARY_TAGGED_STRUCT_IMPL(TYPE_NAME, CATEGORY, TYPE_ID, MAJOR, MINOR) \
+    typedef TYPE_NAME ThisType; \
+    typedef ::slang::PrimaryTag Tag; \
+    static const ::slang::StructTag kStructTag = SLANG_MAKE_PRIMARY_STRUCT_TAG(CATEGORY, TYPE_ID, MAJOR, MINOR); \
+    slang::StructTag structTag = kStructTag; \
+    slang::StructSize structSize = slang::StructSize(sizeof(ThisType)); \
+    const slang::StructTag** exts = nullptr; \
+    int32_t extsCount = 0; 
+
+#define SLANG_API_TYPE_ENUM(x) x,
+
+struct TaggedStructBase
+{
+    StructTag structTag;            ///< Identity for the type
+    StructSize structSize;             ///< Size of the type in bytes
+};
+
+/* Layout for a PrimaryTaggedStruct */
+struct PrimaryTaggedStruct
+{
+    StructTag structTag;                ///< Identity for the type
+    StructSize structSize;              ///< Size of the type in bytes
+    const StructTag** exts;             ///< Extensions - StructTag type should be the first member of Extension structs
+    int32_t extsCount = 0;              ///< The number of extensions
+};
+
+/* Layout for a Extension TaggedStruct */
+struct ExtensionTaggedStruct 
+{
+    StructTag structTag;                ///< Identity for the type
+    StructSize structSize;              ///< Size of the type in bytes
+};
+
+// Enumerate all of the types that want ABI handling as part of Slang API.
+//
+// NOTE! That care is needed using the enumeration, because we want IDs to remain stable.
+// New types should be added to the end. Removed types should become Depreciated_Name (say).
+//
+// NOTE! The names must match the type name for an active type exactly for the SLANG_PRIMARY_TAGGED_STRUCT and SLANG_EXTENSION_TAGGED_STRUCT
+// macros to work.
+#define SLANG_TAGGED_STRUCTS(x) \
+    x(TargetDesc) \
+    x(SessionDesc)
```

Comment:
Style note: I always like to have multiline macro definitions like this terminate each content line with `\` and then have a stand-along line to mark the actual end of the macro, e.g.:

```
#define SLANG_TAGGED_STRUCTS(x) \
    x(TargetDesc) \
    x(SessionDesc) \
    /* end */
```

Formatting things that avoids spurious diffs to the existing lines when adding new cases, and helps avoid the kind of editing error where somebody adds a new case *with* a trailing `\` and accidentally makes the macro definition include a subsequent code line by accident.